src/cadenza.js

```diff
@@ -502,6 +572,42 @@ function validGeometryType(value) {
   ].includes(value);
 }
 
+/** @param {string} [value] */
+function validMediaType(value) {
+  return value === CSV || value === MS_EXCEL_2007;
+}
+
+/** @param {Promise<Response>} responsePromise
+ * @param {object} [options] - Options
+ * @param {string} [options.filename] - The media type to use for the data
+ * */
+async function download(responsePromise, { filename } = {}) {
+  const res = await responsePromise;
+  assert(res.ok, 'download failed');
+  filename = getFilename(res, filename);
+
+  const a = document.createElement('a');
+  a.href = URL.createObjectURL(new Blob([await res.arrayBuffer()]));
+  a.hidden = true;
+  a.download = filename ?? '';
+
+  document.body.appendChild(a);
+  a.click();
+  a.remove();
+}
+
+/** @param {Response} res
+ * @param {string} [filename]
+ * */
+function getFilename(res, filename) {
+  const nameFromHeader = res.headers
+    .get('Content-Disposition')
+    ?.match(/filename[^;=\n]*=((['"]).*?\2|[^;\n]*)/)?.[1]
+    ?.replace(/['"]/g, '');
+  const format = '.' + nameFromHeader?.split('.').at(-1);
+  return filename ? filename + format : nameFromHeader;
+}
+
```

Comment:
imo `format` shouldn't contain a leading dot. Also, you're using the term "filename" here for two different things: the basename of the file and the complete filename.

suggestion:

```
function getFileName (response, baseName) {
  const fileNameFromHeader = …
  if (baseName) {
    const format = fileNameFromHeader?.split('.').at(-1);
    return `${baseName}.${format}`
  }
  return fileNameFromHeader;
}
```

Reply:
I hope we don't need this, see my comment in downloadData().

Reply:
This code is no longer needed because it's handled by the server.